File: Platform/Workspace/Typescript/src/allors/framework/database/data/Contains.ts
import { PropertyType } from '../../meta';

import { ISessionObject } from './../../workspace/SessionObject';
import { Predicate } from './Predicate';

export class Contains implements Predicate {
  public propertyType: PropertyType;
  public object: ISessionObject | string;

  constructor(fields?: Partial<Contains> | PropertyType, object?: ISessionObject | string) {
    if ((fields as PropertyType).objectType) {
      this.propertyType = fields as PropertyType;
      this.object = object;
    } else {
      Object.assign(this, fields);
    }
  }

  public toJSON(): any {
    return {
      kind: 'Contains',
      propertyType: this.propertyType.id,
      object: this.object ? (this.object as ISessionObject).id ? (this.object as ISessionObject).id : this.object : undefined,
    };
  }
}

File: Core/Workspace/CSharp/Domain/Export/Core/Data/Contains.cs
//------------------------------------------------------------------------------------------------- 
// <copyright file="Contains.cs" company="Allors bvba">
// Copyright 2002-2017 Allors bvba.
// 
// Dual Licensed under
//   a) the Lesser General Public Licence v3 (LGPL)
//   b) the Allors License
// 
// The LGPL License is included in the file lgpl.txt.
// The Allors License is an addendum to your contract.
// 
// Allors Platform is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// For more information visit http://www.allors.com/legal
// </copyright>
//-------------------------------------------------------------------------------------------------

using Allors.Workspace.Meta;

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;
    using Allors.Protocol.Data;

    public class Contains : IPropertyPredicate
    {
        public Contains(IPropertyType propertyType = null) => this.PropertyType = propertyType;

        public IPropertyType PropertyType { get; set; }

        public SessionObject Object { get; set; }

        public string Parameter { get; set; }

        public Predicate ToJson()
        {
            return new Predicate
            {
                Kind = PredicateKind.Contains,
                PropertyType = this.PropertyType?.Id,
                Object = this.Object?.Id.ToString(),
                Parameter = this.Parameter
            };
        }

        bool IPredicate.ShouldTreeShake(IReadOnlyDictionary<string, object> arguments)
        {
            return ((IPredicate)this).HasMissingArguments(arguments);
        }

        bool IPredicate.HasMissingArguments(IReadOnlyDictionary<string, object> arguments)
        {
            return this.Parameter != null && (arguments == null || !arguments.ContainsKey(this.Parameter));
        }
    }
}